### PR TITLE
Add MSVC support for Response Files 

### PR DESCRIPTION
**Pull request by temportalflux**
# Goal

Add support for response/command files in MSVC as defined by the [MSVC](https://docs.microsoft.com/en-us/cpp/build/reference/at-specify-a-compiler-response-file) and [Visual Studio MSBuild](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-response-files?view=vs-2019) docs.

# Related Issues

History of Response File support in MSVC implementation

- #107 mentions an interest in supporting response files for MSVC
- The above discussion inspired #192 in 2017, it was abandoned by the author an elongated feedback cycle.
- That PR was followed by another PR (#378), which was also abandoned after the author was redirected to #192.
- #1082 was opened in Dec 2021, which again requested support for response files in MSVC - it has not received any replies. This issue is still open.
- #1183 repeated the same request in June 2022, and a comment was added referencing all of the above tickets. This issue is still open.
- I started discussion #1594 in preparation for this PR.

# Implementation Details 

This change adds an iterator layer between the command-line argument iterator and the `ArgIter`, which used to compare arguments against the supported flags/options. This new layer determines if an option is a response-file directive (`@file`), and if it is, reads the options from the file before continuing to iterate over the command-line args. This requires an additional file-parsing iterating (`SplitArgs`) to split the file contents into arguments in a way which is consistent with the file format.

The `encoding` crate is used to read utf-8 (default encoding in rust) & utf-16 (big and little endian) encodings. The latter is used by `MSBuild` when generating response files.

Please see the `Response Files` doc in this PR for an overview of how response files are used in MSVC, what the expected format is, how they differ from GCC, and how they are implemented for both GCC and MSVC in SCCache.

# Flags

This pull request also includes a handful of changes/additions to the MSVC flags. These flags could be separated into a separate PR, but given how small the changes are, they are currently bundled in with the MSVC changes. Our environment requires these flags, hence the need to have them supported.

## Updated Flags

- `/FC` ([previously marked as TooHardFlag](https://github.com/proletariatgames/sccache/commit/bf6fb5de6f566d20321bf4a47714398250975d03)) changed to `PassThrough` because it only affects how the compiler prints debug information. [flag docs](https://learn.microsoft.com/en-us/cpp/build/reference/fc-full-path-of-source-code-file-in-diagnostics?view=msvc-170)

## Added Flags

Flags which allow users to pass "disable" feature flags which do not affect cachability:

- `/Gm-`, disable minimal rebuilds. [flag docs](https://learn.microsoft.com/en-us/cpp/build/reference/gm-enable-minimal-rebuild?view=msvc-170)
- `/WX-`, disable "treat all compiler warnings as errors". [flag docs](https://learn.microsoft.com/en-us/cpp/build/reference/compiler-option-warning-level?view=msvc-170)
- `/openmp-`, disable the openmp directive. [flag docs](https://learn.microsoft.com/en-us/cpp/build/reference/openmp-enable-openmp-2-0-support?view=msvc-170)

Flags which enable features:

- `/permissive`, allows code which doesn't conform to standard, but doesnt change what needs to be compiled, just how the compiler does so. This is major for some legacy projects which are not up-to-date to standard conformance. [flag docs](https://learn.microsoft.com/en-us/cpp/build/reference/permissive-standards-conformance?view=msvc-170)

### CODE REVIEW

**Comment by Xuanwo** [comment]
**suggestion**: Can you add a basic trst for this feature?

**Comment by temportalflux** [comment]
> **suggestion**: Can you add a basic test for this feature?

What do you mean by "basic test"? There are a number of tests in `compiler/msvc.rs`, including [`test_responsefile_absolute_path`](https://github.com/mozilla/sccache/pull/1618/files#diff-da7fa688de141fae3e6e48ed82ee4e6f6df5925cdbc0299d5a4e79b6f8775b3aR1826), which is the minimal viable successful test (the response file exists and contains arguments). Is there a more basic/minimal test you'd expect to see?

**Comment by Xuanwo** [comment]
I mean add a basic integration test like what we do in `tests`

**Comment by temportalflux** [comment]
Hm yeah I can look into that. I honestly didn't even notice the integration tests, and I haven't yet found any documentation on how those tests are added to or maintained. I'll look through the commit history to see if I can find a decent template/example.

**Review by drahnr — COMMENTED** [review]
First superficial pass, this is the quality we strive for ❤️ thank you for the time and effort put in.

Two questions: Is there a recursion limit on eval? Is that spec'd? And could you write an integration test as @Xuanwo asked?

The only thing after that is appeasing CI

**Comment by sylvestre** [comment]
> I'll look through the commit history to see if I can find a decent template/example.

I love docs but here, copying the current model is easy enough:
https://github.com/mozilla/sccache/blob/main/.github/workflows/integration-tests.yml#L30-L76

**Comment by temportalflux** [comment]
> Is there a recursion limit on eval? Is that spec'd?

Yes there is a limit! I noted it in the `docs/ResponseFiles.md` doc; `A response file cannot contain additional `@file` options, they are not recursive.`. That comes straight out of the [MSVC docs](https://learn.microsoft.com/en-us/cpp/build/reference/at-specify-a-compiler-response-file?view=msvc-170): `It is not possible to specify the @ option from within a response file. That is, a response file cannot embed another response file.`

---

Integration testing wise, I'm looking into that now. Seems like there is little to no obvious docs on adding one, but it looks like the pattern is adding a new function call to `run_sccache_command_tests` under the correct compiler (`cl.exe` for msvc) which runs the integration test (like `test_msvc_deps`).

---

Looks like CI is failing because clippy says `let...else` statements are unstable, even though they [were stabilized recently](https://github.com/rust-lang/rust/issues/87335). I can change the syntax on those two statements if folks prefer to not use `let...else`.

**Comment by temportalflux** [comment]
Added a fairly minimal integration test. Please nitpick it because how the integration tests work in this repo is still somewhat evading me.

File: src/compiler/msvc.rs
Comment (drahnr) [inline src/compiler/msvc.rs:1232]:
```suggestion
}

```

File: src/compiler/msvc.rs
Comment (drahnr) [reply src/compiler/msvc.rs:1222]: suggestion: rename to `SplitMsvcCompilerArgs`

**Review by drahnr — APPROVED** [review]
A few small nits, generally not a a fan of manually parsing file formats, but in this case it's ok.

Thank you for you contribution! I really like the documentation you created in this PR!

File: src/compiler/msvc.rs
Comment (temportalflux) [reply src/compiler/msvc.rs]: Good call; I renamed it to `SplitMsvcResponseFileArgs` to make it clear that its used to split the arguments in a response file. Also swapped the tuple-struct for a named field `file_content`, and made a From impl to create the struct.

**Comment by temportalflux** [comment]
> A few small nits, generally not a a fan of manually parsing file formats, but in this case it's ok.
> 
> Thank you for you contribution! I really like the documentation you created in this PR!

Thanks for the review, and I'm glad the documentation is helpful! I've addressed the nits and squashed all subsequent changes into the original 2 commits after rebasing onto latest `main`.

**Comment by temportalflux** [comment]
@drahnr @sylvestre what does the timeline on getting this merged look like? Any chance it can make it in for the next pre-release or full release (0.4.0)?

**Comment by sylvestre** [comment]
I still would like to see an end to end integration test in
https://github.com/mozilla/sccache/blob/main/.github/workflows/integration-tests.yml if it makes sense

**Review by Xuanwo — APPROVED** [review]

**Comment by temportalflux** [comment]
> I still would like to see an end to end integration test in https://github.com/mozilla/sccache/blob/main/.github/workflows/integration-tests.yml if it makes sense

I'm not sure I fully follow. Are you saying that there should be a new job in the integration tests workflow which runs `cargo test` on the `windows-2019` os against the `x86_64-pc-windows-msvc` to test msvc?

**Comment by sylvestre** [comment]
Nope, building a fake project that uses MSVC / response files

**Comment by temportalflux** [comment]
The above force push is just rebasing on latest so its easy to merge when that time comes

**Review by drahnr — APPROVED** [review]

**Comment by sylvestre** [comment]
`ilammy/msvc-dev-cmd@v1.4.1 is not allowed to be used in mozilla/sccache.`

is there an alternative ? 
(i can ask to the Mozilla admin to enable it but it might take a few days/weeks)

**Comment by temportalflux** [comment]
Given that there are no other integration tests comparable new msvc one (they are all testing integrations with outside tools, not the underlying compilers), and I'd prefer to be efficient about work spent while minimizing delay time, I think the best route forward is:
1. Separate the integration test job into its own PR, whiches uses `ilammy/msvc-dev-cmd` since it seems to be the standard
2. Merge this PR w/o the integration test
3. Mark the integration test PR as blocked until `ilammy/msvc-dev-cmd` is approved

**Comment by temportalflux** [comment]
@sylvestre any thoughts on the msvc gh-action dependency & split-the-PR solution?

**Comment by drahnr** [comment]
I fully agree, we don't need to mention it in Para, wirst case it's just as broken as before

**Comment by sylvestre** [comment]
Please give me a week to see if we can get it enabled
AFAIK, there isn't any rush.

**Comment by sylvestre** [comment]
I think we enabled the github action. Could you please rebase it to retrigger the CI? for some reasons, I can't do it on github on this PR ?!

**Comment by temportalflux** [comment]
Rebased and pushed!

**Comment by temportalflux** [comment]
Whats going on with the `freebsd` check?

**Comment by sylvestre** [comment]
Don't bother. It is quite fragile.

**Comment by temportalflux** [comment]
Looks like it was only failing the integration test because the version allowed was `@v1` and the PR was using the more specific `@v1.4.1` for the `msvc-dev-cmd`. Squashed a change into the last commit which changed the version from `1.4.1` to `1` and hopefully it should be allowed now.

**Comment by sylvestre** [comment]
yeah, i asked if we could fix that on the admin side

**Comment by sylvestre** [comment]
Looks like it worked this time:
```
Compile requests                      2
Compile requests executed             2
Cache hits                            1
Cache hits (C/C++)                    1
Cache misses                          1
Cache misses (C/C++)                  1
```

well done :)

**Comment by assarbad** [comment]
> EDIT: A coworker informed me that `https://github.com/microsoft/setup-msbuild` exists, but after attempting to use it I found that it only sets up msbuild, not the msvc compiler `cl.exe`. So no change to the above comment from this discovery, just more context.

Shameless plug: https://github.com/assarbad/scripts/blob/master/setvcvars.cmd

**Pull request merged**
